# Duplicated, Missing or Irrelevant Data 

### DATA EXPLORATION:
    
 - Duplicated data : ROWS / COLUMNS 
 - Irrelevant data
 - Missing values

In [89]:
import pandas as pd
import numpy as np

In [90]:
df = pd.read_csv("data/attacks.csv", encoding ='ISO-8859-1')

In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

In [92]:
df.shape

(25723, 24)

In [93]:
#format lables for improve analisis:
df.columns = [i.lower().replace(" ", "_").strip() for i in df.columns]

In [94]:
df.shape

(25723, 24)

In [95]:
df.columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex_', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species_', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

In [96]:
df

case_number         date    year        type    country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  area                        location     activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  name sex_  ...         species_  \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
25718              NaN  NaN  ...              NaN   
25719              NaN  NaN  ...              NaN   
25720              NaN  NaN  ...              NaN   
25721              NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               investigator_or_source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
25718                             NaN                       NaN   
25719                             NaN                       NaN   
25720                             NaN                       NaN   
25721                             NaN                       NaN   
25722                             NaN                       NaN   

                                            href_formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
25718                                                NaN   
25719                                                NaN   
25720                                                NaN   
25721                                                NaN   
25722                                                NaN   

                                                    href case_number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

## Analize duplicate rows

In [97]:
df.drop_duplicates().shape

(6312, 24)

In [98]:
#check how many duplicate rows there are:
duplicates = df.shape[0] - df.drop_duplicates().shape[0]

print(f'There is {duplicates} duplicate rows')

There is 19411 duplicate rows


In [99]:
# ESTO PONERLO EN TRANFORMATIONS 
df.drop_duplicates(inplace = True) #ahora lo aplico 

In [100]:
df

case_number         date    year        type    country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
6307            0          NaN     NaN         NaN        NaN   
6308            0          NaN     NaN         NaN        NaN   
6309            0          NaN     NaN         NaN        NaN   
8702          NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  area                        location     activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
6307               NaN                             NaN          NaN   
6308               NaN                             NaN          NaN   
6309               NaN                             NaN          NaN   
8702               NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  name sex_  ...         species_  \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
6307               NaN  NaN  ...              NaN   
6308               NaN  NaN  ...              NaN   
6309               NaN  NaN  ...              NaN   
8702               NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               investigator_or_source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
6307                              NaN                       NaN   
6308                              NaN                       NaN   
6309                              NaN                       NaN   
8702                              NaN                       NaN   
25722                             NaN                       NaN   

                                            href_formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
6307                                                 NaN   
6308                                                 NaN   
6309                                                 NaN   
8702                                                 NaN   
25722                                                NaN   

                                                    href case_number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

### Analize duplicate columns

In [101]:
def cuantos_iguales(col1,col2):
    
    """This functions return the % of duplicated items 
        between 2 given columns"""
    
    #1.Generates a list of True/False items if duplicate/unique
    a = df[col1] == df[col2] 
    
    #2.Set two counters(duplicate/unique)
    count_dupli = 0
    count_unique = 0
    
    #3.Iterates de true/false list to count for duplicates/uniques values
    for i in a:
        if i == True:
            count_dupli +=1
        else:
            count_unique += 1
            
    #4.Calculates % of duplicity:     
    total_values = count_dupli + count_unique
            
    dupli_percent = int((count_dupli / total_values)*100)

    return print(f' columns "{col1}" and "{col2}" are {dupli_percent}% equal')

In [102]:
cuantos_iguales("case_number.1","case_number")

 columns "case_number.1" and "case_number" are 99% equal


In [103]:
cuantos_iguales("case_number.2","case_number")

 columns "case_number.2" and "case_number" are 99% equal


In [104]:
cuantos_iguales("case_number.2","case_number.1")

 columns "case_number.2" and "case_number.1" are 99% equal


- As case_number, case_number.1 and case number.2 are 99% equal
- drop 2 of them we keep case_number

In [105]:
cuantos_iguales("href","href_formula")

 columns "href" and "href_formula" are 98% equal


- we can drop one of them  --> drop href_formula

In [106]:
#data in "href" is the link to the document in pdf

In [107]:
#reggex patern to match de pdf document in the link:
pattern = r'(?<=pdf_directory\/)([^\/]+\.pdf)'

In [108]:
df['href_pdf'] = df["href"].str.extract(pattern)

In [109]:
cuantos_iguales("pdf","href_pdf")

 columns "pdf" and "href_pdf" are 98% equal


- info in href is the same than in pdf --> drop href

In [110]:
df[["case_number","pdf"]]

case_number                       pdf
0      2018.06.25      2018.06.25-Wolfe.pdf
1      2018.06.18    2018.06.18-McNeely.pdf
2      2018.06.09     2018.06.09-Denges.pdf
3      2018.06.08  2018.06.08-Arrawarra.pdf
4      2018.06.04      2018.06.04-Ramos.pdf
...           ...                       ...
6307            0                       NaN
6308            0                       NaN
6309            0                       NaN
8702          NaN                       NaN
25722          xx                       NaN

[6312 rows x 2 columns]

In [111]:
for i in df["pdf"]:
    print(i)

2018.06.25-Wolfe.pdf
2018.06.18-McNeely.pdf
2018.06.09-Denges.pdf
2018.06.08-Arrawarra.pdf
2018.06.04-Ramos.pdf
2018.06.03.b-FlatRock.pdf
2018.06.03.a-daSilva.pdf
2018.05.27-Ponce.pdf
2018.05.26.b-High.pdf
2018.05.26.a-DaytonaBeach.pdf
2018.05.24-CairnsAquarium.pdf
2018.05.21-deBoer.pdf
2018.05.13.b-Turrell.pdf
2018.05.13.a-Berryman.pdf
2018.05.00-Brunning.pdf
2018.05.12-EastLondon.pdf
2018.05.09-SharpesBeach.pdf
2018.04.30.R-Thailand.pdf
2018.04.28.b-CostaRica.pdf
2018.04.28.a-Murray.pdf
2018.04.25.b-Lee.pdf
2018.04.25.a-Bruce.pdf
2018.04.24-Burch.pdf
2018.04.23-Maldives.pdf
2018.04.22-PlettenbergBay.pdf
2018.04.19-McWilliams.pdf
2018.04.15.d-Thailand.pdf
2018.04.15.c-deMelo.pdf
2018.04.15.b-Longrass.pdf
2018.04.15.a-Travaglini.pdf
2018.04.14-Rowan.pff
2018.04.10.R-Paz.pdf
2018.04.09-Magenta.pdf
2018.04.05-McConnell.pdf
2018.04.03-StFrancisBay.pdf
2018.03.31-Hawaii.pdf
2018.03.14-WatermansBay.pdf
2018.03.09.b-Crossland.pdf
2018.03.09.a-Weight.pdf
2018.02.24-Polly.pdf
2018.02.23-Shurap

In [112]:
# to extract dd.mm.aa or ND-dddd
pattern_2 = r"(\d{4}\.\d{2}\.\d{2}|ND-\d{4})"

In [113]:
df['case_pdf'] = df["pdf"].str.extract(pattern_2)

In [114]:
cuantos_iguales("case_pdf","case_number")

 columns "case_pdf" and "case_number" are 64% equal


- We can't consider that is duplicate values with 64%
- Information related to dates and locations

### Resume: Duplicate values :

- Drop duplicare rows --> 19411 duplicate rows droped
- Drop columns "case_number.1" and "case_number.2"
- Drop column "href_formula"
- Drop colum "herf" as relevant info is duplicate with "pdf"

## Missig Rows/Columns

In [115]:
df.shape

(6312, 26)

 - "unnamed:_22" and "unnamed:_23" have just 1 and 2 non-null values

In [116]:
df["unnamed:_22"].unique() #Not relevant values

array([nan, 'stopped here'], dtype=object)

In [117]:
df["unnamed:_23"].unique() #Not relevant values 

array([nan, 'Teramo', 'change filename'], dtype=object)

In [118]:
pd.isnan(df["age"]) ## justificar que la mitad de los datos estan vacios y fuera 

AttributeError: module 'pandas' has no attribute 'isnan'

In [119]:
df.dropna(axis = 0, thresh=18).shape[0] #OJO NO ESTA BIEN 

6267

In [120]:
rows_drop = df.shape[0] - df.dropna(axis = 0, thresh=18).shape[0]

data_drop = (rows_drop/df.shape[0])*100

print(f'{int(data_drop)}%')

0%


 - tresh = 18 --> 18 non-Nan

### Resume missing values:

- Drop "unnamed:_23" and "unnamed:_22" 
`df.drop(["Unnamed: 22","Unnamed: 23"], axis=1, inplace=True`
- "age" colum has almost 50% values Nan --> not acurated data 
- Drop rows to keep those wo have 80% completed values. Reduce 2% dat

## Irrelevant Data

1.  We are going to exclude those columns that we wont be consider for the analisis:
    - investigator_or_source"
    - original order
    - name 
    - values refering time/dates:
        - case_number
        - date
        - year
        - time
        - pdf
    - values refering location:
        - country
        - area
        - location

# Tranform:

In [121]:
import pandas as pd
import numpy as np

df = pd.read_csv("data/attacks.csv", encoding ='ISO-8859-1')

In [122]:
df.columns = [i.lower().replace(" ", "_").strip() for i in df.columns]

In [123]:
#Duplicated values:

df.drop_duplicates(inplace=True)

df.drop("case_number.1", axis=1, inplace=True)
df.drop("case_number.2", axis=1, inplace=True)
df.drop("href_formula", axis=1, inplace=True)
df.drop("href", axis=1, inplace=True)

In [124]:
#Missing data(Nan):

df.drop("unnamed:_23", axis=1, inplace=True)
df.drop("unnamed:_22", axis=1, inplace=True)
df.drop("age", axis=1, inplace=True)
df.dropna(axis = 0, thresh=12, inplace=True) ##ojo con esto revisar arriva

In [125]:
#Irrelevant Data:

df.drop("investigator_or_source", axis=1, inplace=True)
df.drop("original_order", axis=1, inplace=True)
df.drop("time", axis=1, inplace=True)
df.drop("name", axis=1, inplace=True)
df.drop("pdf", axis=1, inplace=True)
df.drop("case_number", axis=1, inplace=True)
df.drop("date", axis=1, inplace=True)
df.drop("year", axis=1, inplace=True)
#df.drop("time", axis=1, inplace=True) -- no la encuentra
df.drop("country", axis=1, inplace=True)
df.drop("area", axis=1, inplace=True)
df.drop("location", axis=1, inplace=True)

In [126]:
df.shape

(6266, 6)

In [127]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6266 entries, 0 to 6301
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   type         6262 non-null   object
 1   activity     5748 non-null   object
 2   sex_         5727 non-null   object
 3   injury       6243 non-null   object
 4   fatal_(y/n)  5740 non-null   object
 5   species_     3455 non-null   object
dtypes: object(6)
memory usage: 342.7+ KB


In [133]:
def export_to_csv(data_frame, file_name):
    """
    Export a DataFrame to a CSV file.

    Parameters:
        - data_frame: pandas DataFrame
        - file_name: str, name of the CSV file (without the extension)
    """
    file_path = f"data/{file_name}.csv"
    
    data_frame.to_csv(file_path, index=False)
    print(f"DataFrame successfully exported to {file_path}")

In [134]:
export_to_csv(df,"export_n1")

DataFrame successfully exported to data/export_n1.csv
